internal/resolver/delegatingresolver/delegatingresolver_test.go

```diff
@@ -0,0 +1,306 @@
+// returns a function to restore the default values.
+func overwrite(hpfe func(req *http.Request) (*url.URL, error)) func() {
+	originalHPFE := HTTPSProxyFromEnvironment
+	HTTPSProxyFromEnvironment = hpfe
+	return func() {
+		HTTPSProxyFromEnvironment = originalHPFE
+	}
+}
+
+// createTestResolverClientConn initializes a test ResolverClientConn and
+// returns it along with channels for resolver state updates and errors.
+func createTestResolverClientConn(t *testing.T) (*testutils.ResolverClientConn, chan resolver.State, chan error) {
+	stateCh := make(chan resolver.State, 1)
+	errCh := make(chan error, 1)
+
+	tcc := &testutils.ResolverClientConn{
+		Logger:       t,
+		UpdateStateF: func(s resolver.State) error { stateCh <- s; return nil },
+		ReportErrorF: func(err error) { errCh <- err },
+	}
+	return tcc, stateCh, errCh
+}
+
+// TestParsedURLForProxyEnv verifies that the parsedURLForProxy function
+// correctly resolves the proxy URL for a given target address.
+func (s) TestParsedURLForProxyEnv(t *testing.T) {
+	// Overwrite the function in the test and restore them in defer.
+	hpfe := func(req *http.Request) (*url.URL, error) {
+		if req.URL.Host == targetTestAddr {
+			return &url.URL{
+				Scheme: "https",
+				Host:   envProxyAddr,
+			}, nil
+		}
+		return nil, nil
+	}
+	defer overwrite(hpfe)()
+
+	// envTestAddr should be handled by ProxyFromEnvironment.
+	got, err := parsedURLForProxy(targetTestAddr)
+	if err != nil {
+		t.Errorf("Unable to get proxy URL : %v\n", err)
+	}
+	if got.Host != envProxyAddr {
+		t.Errorf("want %v, got %v", envProxyAddr, got)
+	}
+}
+
+// TestDelegatingResolverNoProxy verifies that the delegating resolver correctly
+// sends the resolved target URI to the ClientConn.
+func (s) TestDelegatingResolverNoProxy(t *testing.T) {
+	mr := manual.NewBuilderWithScheme("test") // Set up a manual resolver to control the address resolution.
+	target := "test:///" + targetTestAddr
+
+	tcc, stateCh, _ := createTestResolverClientConn(t)
+	// Create a delegating resolver with no proxy configuration
+	dr, err := New(resolver.Target{URL: *testutils.MustParseURL(target)}, tcc, resolver.BuildOptions{}, mr, false)
+	if err != nil || dr == nil {
+		t.Fatalf("Failed to create delegating resolver: %v", err)
+	}
+
+	// Update the manual resolver with a test address.
+	mr.UpdateState(resolver.State{
+		Addresses:     []resolver.Address{{Addr: resolvedTargetTestAddr}},
+		ServiceConfig: &serviceconfig.ParseResult{},
+	})
+
+	// Verify that the delegating resolver outputs the same address.
+	expectedState := resolver.State{
+		Addresses:     []resolver.Address{{Addr: resolvedTargetTestAddr}},
+		ServiceConfig: &serviceconfig.ParseResult{},
+	}
+
+	if state := <-stateCh; len(state.Addresses) != 1 || !cmp.Equal(expectedState, state) {
+		t.Fatalf("Unexpected state from delegating resolver: %v, want %v", state, expectedState)
+	}
+}
+
+// setupDNS unregisters the DNS resolver and registers a manual resolver for the
```

Comment:
There is no real way to *unregister* a resolver. What we are doing instead is to register a new one for the same name thereby overwriting the earlier registration. That needs to be clear in this comment.

Because FYI, we used to have ways to unregister things in the past. We got rid of those for the most part I think. We still have a way to unregister a balancer for testing purposes.